gpt_index/data_structs/data_structs_v2.py

```diff
@@ -0,0 +1,379 @@
+"""Data structures v2.
+
+Nodes are decoupled from the indices.
+
+"""
+
+import random
+import sys
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from typing import Dict, Generic, List, Optional, Set, Tuple, TypeVar
+
+from dataclasses_json import DataClassJsonMixin
+
+from gpt_index.data_structs.data_structs import IndexGraph as V0IndexGraph
+from gpt_index.data_structs.data_structs import IndexStruct, Node
+from gpt_index.data_structs.struct_type import IndexStructType
+from gpt_index.schema import BaseDocument
+
+# OIS = TypeVar("IS", bound=IndexStruct)
+
+
+@dataclass
+class V2IndexStruct(IndexStruct, DataClassJsonMixin):
+    """A base data struct for a LlamaIndex."""
+
+    # # @abstractmethod
+    # @classmethod
+    # def from_v0_struct(cls, v0_struct: OIS) -> "V2IndexStruct":
+    #     """Convert from v0 struct."""
+    #     raise NotImplementedError()
+
+
+@dataclass
+class IndexGraph(V2IndexStruct):
+    """A graph representing the tree-structured index."""
+
+    # mapping from index in tree to Node doc id.
+    all_nodes: Dict[int, str] = field(default_factory=dict)
```

Comment:
Let me type out my understanding of the change, let me know if I'm on the same page.

Previously, `DocumentStore` only contains raw `Document` (i.e. parsed from user input). We were saving Nodes as part of the `IndexStruct`. 

Now, we explicitly store `Node` as part of `DocumentStore` and remove direct reference to `Node` in `V2IndexStruct`. So in terms of serialized representation, index struct only stores the logical relationship between nodes, but not the actual node data structures.